### DISCUSSION TITLE
The question of ZED, and bash

### DISCUSSION THREAD

**Discussion by lundman** [openzfsonwindows/openzfs#409] [Development]
Watching the OpenZFS User Summit, it reminded me that we should be better at alerting the user to faults, which we currently do not do.

So it is time to port `zed` over, to handle that for us. The biggest thing there is that all the fault scripts are written in bash. 

It is probably possible to get bash scripts to run, but it wouldn't be pretty and have somewhat large install requirements. Either cygwin, or Ming, or perhaps WSL. 

ChatGPT did point out that perhaps using Powershell (PS1) is a better fit, since I can link `zed` against PowerShell 5.1, which comes with Windows 7 and up.  (There is PowerShell 7, but that is a separate install, and honestly the zed scripts probably do not need any new features in there). 

It would mean our own copies of the bash scripts for Windows. These scripts aren't particularly complicated, they generally just pull out the ENV vars, pop up notifications or send emails.

Thoughts?

**Comment by BellCubeDev** [comment]
I've rewritten a couple of bash scripts in PowerShell in my time and, honestly, it's something AI is great at—so long as you check its work, of course. I did mine with GitHub Copilot, but ChatGPT should also do a fine job at the bulk of translating.

**Reply by lundman** [reply]
There is the initial bulk, but we would also sign up for a continued updating whenever the originals are changed. Which isn't too often luckily

**Reply by BellCubeDev** [reply]
Yeah, that's likely the biggest issue. For this project, I think your best bet would be something like Snapshot Testing (where a value is stored in Git and checked against in every test run) to make sure you keep the PowerShell scripts up to date.

It'd be simple enough to store the hash of each Bash script in Git and then add some tests to validate that the hashes match the script that's currently checked out. Once you update the corresponding PowerShell script, you can update the hash.

**Reply by BellCubeDev** [reply]
Last time I did something like this, I was working on a SAAS project where I used a "Backend as a Service" solution called Supabase. I wanted to make sure my repo stayed up to date with the latest Supabase release, so I added a test to make sure the Docker images the project referenced matched the latest Supabase release. This case should be a bit easier since you wouldn't have to download anything; just some good old-fashioned file hashing.

**Comment by lundman** [comment]
Well, real basic port exists and example of script structure. Still needs cleaning up to be useful.

![Screenshot 2024-10-28 at 17 01 36 (2)](https://github.com/user-attachments/assets/4f35d865-4da9-49c9-a6b9-fc6a06a3c18c)


https://github.com/openzfsonwindows/openzfs/tree/zed/cmd/zed/os/windows/zed.d
[zed.zip](https://github.com/user-attachments/files/17540574/zed.zip)
`zed.exe -f -vvv -d cmd/zed/os/windows/zed.d -p zed.txt -s zed.status -F`

**Comment by sskras** [comment]
What intro / HOWTO would anyone recommended to learn about setting up and using ZED on a genuine platform? (I think it's Linux)

Found a few pages:
https://hackmd.io/@bilalkhan/HknZMN_-o
https://rair.dev/zfs-smart-ntfy/#zfs-notifications-via-zed
https://utcc.utoronto.ca/~cks/space/blog/linux/ZFSZEDPraise

The first page looks pretty simple. The second one seems to be pretty complicated (and tied to the **ntfy** software). The 3rd one contains some nice insights, but doesn't seem to be complete.

**Reply by lundman** [reply]
The idea with zed is it is generally installed when you ... install.. ZFS, and runs in the background. The defaults should be sane enough, so that if something breaks, it will tell you. You might need to set some values, like email address to send to, should it pop up alert notifications.. those sort of yes/no options. But generally for most users it does the monitoring for them, and nobody needs to play with it. For the advanced types, it is flexible enough that you can make up something tailored to your situation. You want slack notification? Add a script for it.. etc.

But it does need a bit of development to get there, which is what we discuss in here. If people wanted to help out, I'm new to powershell but chatgpt has been excellent.

**Comment by lundman** [comment]
Outline of work:

zed uses syslog on Unix to store events permanently, so you can look up what went wrong, and when. Windows version needs to do something similar, which probably means plugging into Event Viewer system. The installer can take care of the Registry node needed, and the zedlets can use Write-EventLog to put messages in there.

We can pop up toast notifications, probably based on the level (critical? warning?)

Following the general layout  of Unix zedlets is a good idea. You can see all-debug.sh and all-debug.ps1 look similar, does the same include of zed.rc, and zed-functions.ps1.

and so on..
